xarray/backends/api.py

```diff
@@ -549,6 +549,25 @@ def to_netcdf(dataset, path=None, mode='w', format=None, group=None,
             engine = _get_default_engine(path)
         path = _normalize_path(path)
 
+    for var in encoding:
+        # This section sets the default _FillValue
+        # to NaNs
+        if '_FillValue' not in encoding[var]:
+            if 'dtype' in encoding[var]:
+                dtype = encoding[var]['dtype']
+            else:
+                dtype = dataset[var].dtype
+            try:
+                # Test whether NaNs are accepted for
+                # the requested dtype:
+                np.asarray(np.nan, dtype=dtype)
```

Comment:
I think it will be more reliable to explicitly check if the dtype is a float or complexfloat type, e.g., use `if np.issubdtype(x.dtype, np.floating) or np.issubdtype(x.dtype, np.complexfloating):`. Xarray uses dtype=object for arrays of strings, and we don't want to use `np.nan` as a fill value for those.